cylc/flow/pathutil.py

```diff
@@ -335,15 +335,37 @@ def remove_empty_parents(
             break
 
 
-def get_next_rundir_number(run_path):
-    """Return the new run number"""
-    run_n_path = os.path.expanduser(os.path.join(run_path, "runN"))
-    try:
+def get_next_rundir_number(run_path: Union[str, Path]) -> int:
+    """Return the next run number for a new install.
+
+    Args:
+        run_path: Top level installed workflow dir
+        (often ``~/cylc-run/workflow``).
+
+    """
+    run_n_path = Path(os.path.expanduser(os.path.join(run_path, "runN")))
+    if run_n_path.exists() and run_n_path.is_symlink():
         old_run_path = os.readlink(run_n_path)
-        last_run_num = re.search(r'(?:run)(\d*$)', old_run_path).group(1)
-        return int(last_run_num) + 1
-    except OSError:
-        return 1
+        # Line below could in theory not return a match group, so mypy objects.
+        # This function unlikely to be called in circumstances where this will
+        # be a problem.
+        last_run_num = re.search(  # type: ignore[union-attr]
+            r'(?:run)(\d*$)', old_run_path
+        ).group(1)
+        last_run_num = int(last_run_num)
+    else:
+        # If the ``runN`` symlink has been removed, get next numbered run from
+        # file names:
+        last_run_num = 0
+        for rundir in Path(run_path).glob('run[0-9]*'):
+            highest_yet = re.search(  # type: ignore[union-attr]
+                r'(?:run)(\d*$)', str(rundir.name)
+            ).group(1)
+            highest_yet = int(highest_yet)
+            if highest_yet > last_run_num:
+                last_run_num = highest_yet
```

Comment:
Could do a comprehension like
```python
last_run_num = max(
    (
        m.group(1) for m in (
            re.match(r'run(\d+)$', i.name) for i in os.scandir()
        ) if m
    ),
    default=0
)
```

Reply:
This would probably be quicker, but IMO to read: Stet?

Reply:
Efficiency concerns aside, the main benefit I see is avoiding the type error in the case of the regex not matching (because the glob `run[0-9]*` can match `run2B`).

Reply:
The glob matches run2B, but surely this just acts as a pre-filter for the regex?